### ISSUE TITLE
AuthenticationException Masks Multiple Errors

### ISSUE THREAD

**Issue by meibenny** [paramiko/paramiko#2091]
Paramiko uses a generic `AuthenticationException` to mask multiple errors / failure modes. 

For example, the errors `authentication failed.`, and `authentication timeout.` are both raised under `AuthenticationException`.

https://github.com/paramiko/paramiko/blob/1824a27c644132e5d46f2294c1e2fa131c523559/paramiko/auth_handler.py#L244
https://github.com/paramiko/paramiko/blob/1824a27c644132e5d46f2294c1e2fa131c523559/paramiko/auth_handler.py#L249

Is there a reason for this design? Has there been any discussion on separating these errors into their own exception classes? 

It would be a huge convenience for me if these errors were formally given their own classes so they can be caught and independently handled, rather than catching a generic `AuthenticationException` and performing string comparisons to determine the actual failure.

**Comment by bskinn** [comment]
Once Python 2 is dropped, it should be possible to rework this with the Python 3 `raise from` syntax. Would that be enough? Or would a refactor/refinement of the exception classes likely also be a good idea?

**Comment by meibenny** [comment]
Yes, it would be enough for me that there's separate classes handling each error. I am neutral towards refactoring the existing exception classes.

My thought was something along the lines of
```
class AuthenticationFailed(AuthenticationException):
    ...

class AuthenticationTimeout(AuthenticationException):
    ...

```

Then `AuthHandler` can raise 
```
raise AuthenticationFailed("Authentication failed.")
```
instead of the existing
```
raise AuthenticationException("Authentication failed.") 
```
